components/script/dom/htmlformelement.rs

```diff
@@ -265,43 +330,23 @@ impl HTMLFormElementMethods for HTMLFormElement {
 
     // https://html.spec.whatwg.org/multipage/#the-form-element%3Adetermine-the-value-of-a-named-property
     fn NamedGetter(&self, name: DOMString) -> Option<RadioNodeListOrElement> {
-        let mut candidates: Vec<DomRoot<Node>> = Vec::new();
+        let window = window_from_node(self);
 
-        let controls = self.controls.borrow();
         // Step 1
-        for child in controls.iter() {
-            if child
-                .downcast::<HTMLElement>()
-                .map_or(false, |c| c.is_listed_element())
-            {
-                if (child.has_attribute(&local_name!("id")) &&
-                    child.get_string_attribute(&local_name!("id")) == name) ||
-                    (child.has_attribute(&local_name!("name")) &&
-                        child.get_string_attribute(&local_name!("name")) == name)
-                {
-                    candidates.push(DomRoot::from_ref(&*child.upcast::<Node>()));
-                }
-            }
-        }
+        let mut candidates =
+            RadioNodeList::new_controls_except_image_inputs(&window, self, name.clone());
+        let mut candidates_length = candidates.Length();
+
         // Step 2
-        if candidates.len() == 0 {
-            for child in controls.iter() {
-                if child.is::<HTMLImageElement>() {
-                    if (child.has_attribute(&local_name!("id")) &&
-                        child.get_string_attribute(&local_name!("id")) == name) ||
-                        (child.has_attribute(&local_name!("name")) &&
-                            child.get_string_attribute(&local_name!("name")) == name)
-                    {
-                        candidates.push(DomRoot::from_ref(&*child.upcast::<Node>()));
-                    }
-                }
-            }
+        if candidates_length == 0 {
+            candidates = RadioNodeList::new_images(&window, self, name.clone());
+            candidates_length = candidates.Length();
```

Comment:
we could use a separate function that just returns if there are zero, 1, or more (returning the element if it's one)